Add reference page for CEL in Kubernetes (#39642)
kubernetes/website: content/en/docs/reference/using-api/cel.md

---

> title: CEL in Kubernetes

Reviewer: nit: Spell out CEL ? What is it and can it be applied to other systems?

Author: Thanks. I've replaced "CEL" with "Common Expression Language" here.

This document links to official CEL documentation but tries to mostly avoid duplicating that information. The Language Overview is maybe a bit redundant, but is contextualized enough that I hope it is justified. The rest of the content is specific to how CEL integrates with Kubernetes.

reviewers:
- jpbetz
- cici37
content_type: concept
weight: 25
min-kubernetes-server-version: 1.25
---

<!-- overview -->

The [Common Expression Language (CEL)](https://github.com/google/cel-go) is used

> in the Kubernetes API to declare validation rules, policy rules and other

Reviewer 2: supernit: prefer the Oxford comma - "policy rules, and"

constraints or conditions.

> CEL expressions are evaluated directly in the API server, making CEL a

Reviewer 3: ```suggestion
CEL expressions are evaluated directly in the
{{< glossary_tooltip text="API server" term_id="kube-apiserver" >}}, making CEL a
```

convenient alternative to out-of-process mechanisms, such as webhooks, for many

> extensibility use cases.

Reviewer 3: ```suggestion
extensibility use cases. Your CEL expressions continue to execute so long as the
control plane's API server component remains available.
```
?

<!-- body -->

> ### Language Overview

Reviewer 3: (nit) we use sentence case in headings
```suggestion
### Language overview
```

> The [CEL language](https://github.com/google/cel-spec/blob/master/doc/langdef.md)

Reviewer 3: I'd link to https://opensource.google.com/projects/cel - how about that?

Reviewer 2: Unfortunately that link will just redirect to the top-level `cel-spec` repo. It's probably better to preserve the link to the language definition.

Reviewer 3: https://opensource.google.com/projects/cel is a more durable link. We can't as easily promise that https://github.com/google/cel-spec/blob/master/doc/langdef.md stays valid.

| `has(self.expired) && self.created + self.ttl < self.expired`                      | Validate that 'expired' date is after a 'create' date plus a 'ttl' duration       |
| `self.health.startsWith('ok')`                                                     | Validate a 'health' string field has the prefix 'ok'                              |
| `self.widgets.exists(w, w.key == 'x' && w.foo < 10)`                               | Validate that the 'foo' property of a listMap item with a key 'x' is less than 10 |
| `type(self) == string ? self == '100%' : self == 1000`                             | Validate an int-or-string field for both the the int and string cases             |
| `self.metadata.name == 'singleton'`                                                | Validate that an object's name matches a specific value (making it a singleton)   |
| `self.set1.all(e, !(e in self.set2))`                                              | Validate that two listSets are disjoint                                           |
| `self.names.size() == self.details.size() && self.names.all(n, n in self.details)` | Validate the 'details' map is keyed by the items in the 'names' listSet           |

> ### Libraries

Reviewer 3: ```suggestion
## Libraries
```

Libraries available to CEL expressions in Kubernetes include:

- CEL standard functions, defined in the [list of standard definitions](https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#list-of-standard-definitions)
- CEL standard [macros](https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#macros)
- CEL [extended string function library](https://pkg.go.dev/github.com/google/cel-go@v0.11.2/ext#Strings)
- Kubernetes [CEL extension library](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#pkg-functions)

> ### Type Checking

Reviewer 3: (nit) we use sentence case in headings
```suggestion
### Type checking
```

Author: Oops, fixing all headings.

CEL is a [gradually typed language](https://github.com/google/cel-spec/blob/master/doc/langdef.md#gradual-type-checking).

Some Kubernetes API fields contain fully type checked CEL expressions. For
example, [CustomResourceDefinitions Validation

> Rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)

Reviewer 3: ```suggestion
Rules](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)
```

are fully type checked.

Some Kubernetes API fields contain partially type checked CEL expressions. A
partially type checked expression is an experessions where some of the variables
are statically typed but others are dynamically typed. For example, in the CEL
expressions of

> [ValidatingAdmissionPolicies](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/)

Reviewer 3: ```suggestion
[ValidatingAdmissionPolicies](/docs/reference/access-authn-authz/validating-admission-policy/)
```

the `request` variable is typed, but the `object` variable is dynamically typed.
As a result, an expression containing `request.namex` would fail type checking
because the `namex` field is not defined. However, `object.namex` would pass
type checking even when the `namex` field is not defined for the resource kinds
that `object` refers to, because `object` is dynamically typed.

The `has()` macro in CEL may be used in CEL expressions to check if a field of a
dynamically typed variable is accessable before attempting to access the field's
value. For example:

```
has(object.namex) ? object.namex == 'special' : request.name == 'special'
```

> ### Type System Integration

Reviewer 3: (nit) we use sentence case in headings
```suggestion
### Type system integration
```

| 'string' with format=byte (base64 encoded)         | bytes                                                                                                                        |
| 'string' with format=date                          | timestamp (google.protobuf.Timestamp)                                                                                        |
| 'string' with format=datetime                      | timestamp (google.protobuf.Timestamp)                                                                                        |
| 'string' with format=duration                      | duration (google.protobuf.Duration)                                                                                          |

xref: [CEL types](https://github.com/google/cel-spec/blob/v0.6.0/doc/langdef.md#values),
[OpenAPI types](https://swagger.io/specification/#data-types),
[Kubernetes Structural Schemas](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema).

Equality on arrays with `x-kubernetes-list-type` of `set` or `map` ignores element
order, i.e. `[1, 2] == [2, 1]`.

> Concatenation on arrays with` x-kubernetes-list-type` use the semantics of the

Reviewer 3: Concatenation on arrays with `x-kubernetes-list-type` use the semantics of the

  are preserved but the values are overwritten by values in `Y` when the key
  sets of `X` and `Y` intersect. Elements in `Y` with non-intersecting keys are
  appended, retaining their partial order.

### Escaping

Only Kubernetes resource property names of the form
`[a-zA-Z_.-/][a-zA-Z0-9_.-/]*` are accessible from CEL. Accessible property
names are escaped according to the following rules when accessed in the
expression:

| escape sequence   | property name equivalent                                                                     |
|-------------------|----------------------------------------------------------------------------------------------|
| `__underscores__` | `__`                                                                                         |
| `__dot__`         | `.`                                                                                          |
| `__dash__`        | `-`                                                                                          |
| `__slash__`       | `/`                                                                                          |

> | `__{keyword}__`   | [CEL RESERVED keyword](https://github.com/google/cel-spec/blob/v0.6.0/doc/langdef.md#syntax) |

Reviewer 3: ```suggestion
| `__{keyword}__`   | [CEL **RESERVED** keyword](https://github.com/google/cel-spec/blob/v0.6.0/doc/langdef.md#syntax) |
```

> Note that CEL RESERVED keywords need to match the exact property name to be escaped
> (e.g. int in the word sprint would not be escaped).

Reviewer 3: ```suggestion
When you escape any of CEL's **RESERVED** keywords you need to match the exact property name
use the underscore escaping
(for example, `int` in the word `sprint` would not be escaped and nor would it need to be).
```

Examples on escaping:

| property name | rule with escaped property name   |
|---------------|-----------------------------------|
| namespace     | `self.__namespace__ > 0`          |
| x-prop        | `self.x__dash__prop > 0`          |
| redact__d     | `self.redact__underscores__d > 0` |

> | string        | `self.startsWith('kube')`         |

Reviewer 3: Not sure I understand this one. Maybe put `namespace`, `x-prop` and `redact__d` in backticks, and string in italics?

I get that you can refer to `__namespace__` like that but string isn't a reserved word and I don't see any escaping ehre.

Author: Problem is that `namespace` is a reserved keyword in the CEL language grammar.

> ### Resource Constraints

Reviewer 3: (nit) we use sentence case in headings
```suggestion
### Resource constraints
```

> To ensure that CEL expressions are safe to be run directly in the API server,

Reviewer 2: ```suggestion
CEL is non-Turing complete and offers a variety of production safety controls to limit execution time. CEL's _resource constraint_ features provide feedback to developers about expression complexity and help protect the API server from excessive resource consumption during evaluation.
```

Reviewer 3: This suggestion isn't acceptable as-is (committing it would yield bad wording) but I like the intent.

> CEL's "Resource Constraints" features are used to prevent CEL evaluation from

Reviewer 3: ```suggestion
CEL's _resource constraint_ features are used to prevent CEL evaluation from
```

> consuming excessive api-server resources.

Reviewer 3: ```suggestion
consuming excessive API server resources.
```

> #### Runtime Cost Budget

Reviewer 3: (nit) we use sentence case in headings
```suggestion
#### Runtime cost budget
```

All CEL expressions evaluated by Kubernetes are constrained by a runtime cost
budget. If the CEL interpreter executes too many instructions, the runtime cost
budget will be exceeded, execution of the expressions will be halted, and an
error will result.

The runtime cost budget is an estimate of actual CPU utilization computed by
incrementing a "cost unit" counter while interpreting a CEL expression. "Cost
units" are independant of system load and hardware. Cost is also deterministic;
for any given CEL expression and input data, evaluation of the expression by
the CEL interpreter will always result in the same cost.

Some Kubernetes resources may define a an additional runtime cost budget that
bounds the execution of multiple expressions. If the sum total of the cost of
expressions exceed the budget, execution of the expressions will be halted, and
an error will result. For example the validation of a custom resource has a
"per-valiation" runtime cost budget for all [Validation
Rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)
evaluated to validate the custom resource.

> #### Estimated Cost limits

Reviewer 3: (nit) we use sentence case in headings
```suggestion
#### Estimated cost limits
```

consuming excessive API server resources.

### Runtime Cost Budget

All CEL expressions evaluated by Kubernetes are constrained by a runtime cost
budget. If the CEL interpreter executes too many instructions, the runtime cost
budget will be exceeded, execution of the expressions will be halted, and an
error will result.

The runtime cost budget is an estimate of actual CPU utilization computed by
incrementing a "cost unit" counter while interpreting a CEL expression. "Cost
units" are independant of system load and hardware. Cost is also deterministic;
for any given CEL expression and input data, evaluation of the expression by
the CEL interpreter will always result in the same cost.

Some Kubernetes resources may define a an additional runtime cost budget that
bounds the execution of multiple expressions. If the sum total of the cost of
expressions exceed the budget, execution of the expressions will be halted, and
an error will result. For example the validation of a custom resource has a
"per-valiation" runtime cost budget for all [Validation
Rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)

> evaluated to validate the custom resource.

Reviewer 3: It would be nice to document whether or not webhook authz mode has an impact on runtime cost (I feel that it should - but if it doesn't then maybe we ought instead to warn users about the combination of CEL uses cases that allow authz checks, and webhook mode authz).

Author: Good catch. Authz mode does impact runtime cost. I'll clarify in this documentation. I also talked with sig-auth and was informed that the authz webhook calls, for the admission request principal, are already cached at this point, so CEL won't be impacted by webhook latency.

Author: Follow up:

The cost for authz checks is very high now at 10% of the expression budget per check. For cases where all the authz checks are via webhook mode, this seems _close_ to what we want.  I'd be okay with moving it 25% or 30% to make it even more restrictive. Thoughts?

I've also added a short warning about webhooks in the "Authorization library" section.

| `redact__d`   | `self.redact__underscores__d > 0` |
| `string`      | `self.startsWith('kube')`         |

## Resource constraints

CEL is non-Turing complete and offers a variety of production safety controls to
limit execution time. CEL's _resource constraint_ features provide feedback to
developers about expression complexity and help protect the API server from
excessive resource consumption during evaluation. CEL's resource constraint
features are used to prevent CEL evaluation from consuming excessive API server
resources.

A key element of the resource constraint features is a _cost unit_ that CEL
defines as a way of tracking CPU utilization. Cost units are independant of
system load and hardware. Cost units are also deterministic; for any given CEL
expression and input data, evaluation of the expression by the CEL interpreter
will always result in the same cost.

Many of CEL's core operations have fixed costs. The simplest operations, such as
comparisons (e.g. `<`) have a cost of 1. Some have a higher fixed cost, for
example list literal declarations have a fixed base cost of 40 cost units.

Calls to functions implemented in native code approximate cost based on the time

> complexty of the operation. For example: operations that use regular

Reviewer 3: ```suggestion
complexity of the operation. For example: operations that use regular
```

| `self.minReplicas <= self.replicas && self.replicas <= self.maxReplicas`           | Validate that the three fields defining replicas are ordered appropriately        |
| `'Available' in self.stateCounts`                                                  | Validate that an entry with the 'Available' key exists in a map                   |
| `(self.list1.size() == 0) != (self.list2.size() == 0)`                             | Validate that one of two lists is non-empty, but not both                         |
| `self.envars.filter(e, e.name = 'MY_ENV').all(e, e.value.matches('^[a-zA-Z]*$')`   | Validate the 'value' field of a listMap entry where key field 'name' is 'MY_ENV'  |
| `has(self.expired) && self.created + self.ttl < self.expired`                      | Validate that 'expired' date is after a 'create' date plus a 'ttl' duration       |
| `self.health.startsWith('ok')`                                                     | Validate a 'health' string field has the prefix 'ok'                              |
| `self.widgets.exists(w, w.key == 'x' && w.foo < 10)`                               | Validate that the 'foo' property of a listMap item with a key 'x' is less than 10 |

> | `type(self) == string ? self == '100%' : self == 1000`                             | Validate an int-or-string field for both the the int and string cases             |

Reviewer 4: Just curiously, why is the in value 1000 instead of 100 ?

Reviewer 3: I think we could use a different value to make it extra clear the numbers aren't related. Eg, check either for `"100%"` or `7`, reject any other value.

Author: I can use random numbers here. I think the use cases I copied this from was implicitly using millis as the unit of the int, but I agree that this example raises eyebrows the way it is currently written. Fixed.

---
title: Common Expression Language in Kubernetes
reviewers:
- jpbetz
- cici37
content_type: concept
weight: 35
min-kubernetes-server-version: 1.25
---

<!-- overview -->

The [Common Expression Language (CEL)](https://github.com/google/cel-go) is used
in the Kubernetes API to declare validation rules, policy rules, and other
constraints or conditions.

CEL expressions are evaluated directly in the
{{< glossary_tooltip text="API server" term_id="kube-apiserver" >}}, making CEL a
convenient alternative to out-of-process mechanisms, such as webhooks, for many
extensibility use cases. Your CEL expressions continue to execute so long as the
control plane's API server component remains available.

<!-- body -->

## Language Overview

The [CEL language](https://github.com/google/cel-spec/blob/master/doc/langdef.md)

> has a straight-forward and unsurprising grammar.

Reviewer: straightforward,
not straight-forward

nit: What does `unsurprising grammar` mean?

Author: Thanks!  I'll rephrase to "syntax that is similar to the expressions in with C, C++, Java,
JavaScript or Go". Suggestions welcome!

The list library also includes `min`, `max` and `sum`. Sum is supported on all
number types as well as the duration type. Min and max are supported on all
comparable types.

`isSorted` is also provided as a convenience function and is supported on all
comparable types.

Examples:

| CEL Expression                                                                     | Purpose                                                   |
|------------------------------------------------------------------------------------|-----------------------------------------------------------|
| `names.isSorted()`                                                                 | Verify that a list of names is kept in alphabetical order |
| `items.map(x, x.weight).sum() == 1.0`                                              | Verify that the "weights" of a list of objects sum to 1.0 |
| `lowPriorities.map(x, x.priority).max() < highPriorities.map(x, x.priority).min()` | Verify that two sets of priorities do not overlap         |
| `names.indexOf('should-be-first') == 1`                                            | Require that the first name in a list if a specific value | 

> See the [List Library godoc](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library#Lists) for more information.

Reviewer 3: ```suggestion
See the [Kubernetes List Library](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library#Lists)
godoc for more information.
```

### Kubernetes Regex Library

In addition to the `matches` function provided by the CEL standard library, the
regex library provides `find` and `findAll`, enabling a much wider range of
regex operations.

Examples:

| CEL Expression                                              | Purpose                                                  |
|-------------------------------------------------------------|----------------------------------------------------------|
| `"abc 123".find('[0-9]*')`                                  | Find the first number in a string                        |
| `"1, 2, 3, 4".findAll('[0-9]*').map(x, int(x)).sum() < 100` | Verify that the numbers in a string sum to less than 100 |

> See the [Regex Library godoc](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library#Regex) for more information.

Reviewer 3: ```suggestion
See the [Kubernetes Regex Library](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library#Regex)
godoc for more information.
```

### Kubernetes URL Library

To make it easier and safer to process URLs, the following functions have been added:

- `isURL(string)` - Check if a string is a valid URL according to the [Go's
  net/url](https://pkg.go.dev/net/url#URL) package. The string must be an
  absolute URL.
- `url(string) URL` Converts a string to a URL or results in an error if the
  string is not a valid URL.

Once parsed via the `url` function, the resulting URL object has `getScheme`,
`getHost`, `getHostname`, `getPort`, `getEscapedPath` and `getQuery` accessors.

Examples:

| CEL Expression                                                  | Purpose                                        |
|-----------------------------------------------------------------|------------------------------------------------|
| `url('https://example.com:80/').getHost()`                      | Get the 'example.com:80' host part of the URL. |
| `url('https://example.com/path with spaces/').getEscapedPath()` | Returns '/path%20with%20spaces/'               |

> See the [URL Library godoc](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library#URLs) for more information.

Reviewer 3: ```suggestion
See the [Kubernetes URL Library](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library#URLs)
godoc for more information.
```

Some Kubernetes API fields contain fully type checked CEL expressions. For
example, [CustomResourceDefinitions Validation
Rules](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)
are fully type checked.

Some Kubernetes API fields contain partially type checked CEL expressions. A
partially type checked expression is an experessions where some of the variables
are statically typed but others are dynamically typed. For example, in the CEL
expressions of
[ValidatingAdmissionPolicies](/docs/reference/access-authn-authz/validating-admission-policy/)
the `request` variable is typed, but the `object` variable is dynamically typed.
As a result, an expression containing `request.namex` would fail type checking
because the `namex` field is not defined. However, `object.namex` would pass
type checking even when the `namex` field is not defined for the resource kinds
that `object` refers to, because `object` is dynamically typed.

The `has()` macro in CEL may be used in CEL expressions to check if a field of a
dynamically typed variable is accessable before attempting to access the field's
value. For example:

> ```

Reviewer 3: You could append `cel` to this line, and then any future syntax highlighting support will pick it up.

Author: Oh nice. Thx!

| 'array                                             | list                                                                                                                         |
| 'array' with x-kubernetes-list-type=map            | list with map based Equality & unique key guarantees                                                                         |
| 'array' with x-kubernetes-list-type=set            | list with set based Equality & unique entry guarantees                                                                       |
| 'boolean'                                          | boolean                                                                                                                      |
| 'number' (all formats)                             | double                                                                                                                       |
| 'integer' (all formats)                            | int (64)                                                                                                                     |

> | <no equivalent>                                    | uint (64)                                                                                                                    |

Reviewer 3: ```suggestion
| _no equivalent_                                    | uint (64)                                                                                                                    |
```

| 'null'                                             | null_type                                                                                                                    |
| 'string'                                           | string                                                                                                                       |
| 'string' with format=byte (base64 encoded)         | bytes                                                                                                                        |
| 'string' with format=date                          | timestamp (google.protobuf.Timestamp)                                                                                        |
| 'string' with format=datetime                      | timestamp (google.protobuf.Timestamp)                                                                                        |
| 'string' with format=duration                      | duration (google.protobuf.Duration)                                                                                          |

> xref: [CEL types](https://github.com/google/cel-spec/blob/v0.6.0/doc/langdef.md#values),

Reviewer 3: ```suggestion
Also see: [CEL types](https://github.com/google/cel-spec/blob/v0.6.0/doc/langdef.md#values),
```

[OpenAPI types](https://swagger.io/specification/#data-types),
[Kubernetes Structural Schemas](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema).

> Equality on arrays with `x-kubernetes-list-type` of `set` or `map` ignores element

Reviewer 3: ```suggestion
Equality comparison for arrays with `x-kubernetes-list-type` of `set` or `map` ignores element
```

> order, i.e. `[1, 2] == [2, 1]`.

Reviewer 3: ```suggestion
order. For example `[1, 2] == [2, 1]` if the arrays represent Kubernetes `set` values.
```

(for example, `int` in the word `sprint` would not be escaped and nor would it need to be).

Examples on escaping:

| property name | rule with escaped property name   |
|---------------|-----------------------------------|
| `namespace`   | `self.__namespace__ > 0`          |
| `x-prop`      | `self.x__dash__prop > 0`          |
| `redact__d`   | `self.redact__underscores__d > 0` |
| `string`      | `self.startsWith('kube')`         |

## Resource Constraints

CEL is non-Turing complete and offers a variety of production safety controls to
limit execution time. CEL's _resource constraint_ features provide feedback to
developers about expression complexity and help protect the API server from
excessive resource consumption during evaluation. CEL's _resource constraint_
features are used to prevent CEL evaluation from consuming excessive API server
resources.

A key element of the _resource constraint_ features is a "cost unit" that CEL
defines as a way of tracking CPU utilization. "Cost units" are independant of
system load and hardware. "Cost units" are also deterministic; for any given CEL
expression and input data, evaluation of the expression by the CEL interpreter

> will always result in the same cost.

Reviewer 3: Use emphasis only for the first use of a jargon-y term: _resource constraint_ and _cost unit_. Subsequent uses don't need emphasis.

Author: Thanks, I've also applied the treatment to _cost unit_.

Many of CEL's core operations have fixed costs. The simplest operations, such as
comparisons (e.g. `<`) have a cost of 1. Some have a higher fixed cost, for
example:

- List literal declarations have a fixed base cost of 40 cost units.
- Authorization checks have a high fixed cost of 100000 cost units, effectively
  limiting the total number of checks for any expression to 10.

More complex functions approximate cost based on the time complexty of the

> operation, for example regex operations such as `match` and `find` are estimated

Reviewer 3: ```suggestion
operation. For example: operations that use regular expressions, such as `match` and `find`,
are estimated
```

> as roughly as `length(regexString)*length(inputString)` to reflect the worst

Reviewer 3: ```suggestion
using an approximation similar to `length(regexString)*length(inputString)`. The approximation
aims to reflect the worst
```

error will result.

The runtime cost budget is an estimate of actual CPU utilization computed by
incrementing a "cost unit" counter while interpreting a CEL expression. "Cost
units" are independant of system load and hardware. Cost is also deterministic;
for any given CEL expression and input data, evaluation of the expression by
the CEL interpreter will always result in the same cost.

Some Kubernetes resources may define a an additional runtime cost budget that
bounds the execution of multiple expressions. If the sum total of the cost of
budget. CEL increments a "cost unit" counter while interpreting a CEL expression
to track runtime cost. If the CEL interpreter executes too many instructions,
the runtime cost budget will be exceeded, execution of the expressions will be
halted, and an error will result.

Some Kubernetes resources define an additional runtime cost budget that bounds
the execution of multiple expressions. If the sum total of the cost of
expressions exceed the budget, execution of the expressions will be halted, and
an error will result. For example the validation of a custom resource has a

> "per-valiation" runtime cost budget for all [Validation

Reviewer 3: ```suggestion
_per-validation_ runtime cost budget for all [Validation
```

---
title: Common Expression Language in Kubernetes
reviewers:
- jpbetz
- cici37
content_type: concept
weight: 35
min-kubernetes-server-version: 1.25
---

<!-- overview -->

The [Common Expression Language (CEL)](https://github.com/google/cel-go) is used
in the Kubernetes API to declare validation rules, policy rules, and other
constraints or conditions.

CEL expressions are evaluated directly in the
{{< glossary_tooltip text="API server" term_id="kube-apiserver" >}}, making CEL a
convenient alternative to out-of-process mechanisms, such as webhooks, for many

> extensibility use cases. Your CEL expressions continue to execute so long as the

Reviewer 5: `as long as`?

Author: AFAIK, both "so long as" and "as long as" are considered idiomatic.

control plane's API server component remains available.

<!-- body -->

## Language Overview

The [CEL
language](https://github.com/google/cel-spec/blob/master/doc/langdef.md) has a
straightforward syntax that is similar to the expressions in with C, C++, Java,
JavaScript and Go.

CEL was designed to be embedded into applications. Each CEL "program" is a
single expression that evaluates to a single value. CEL expressions are
typically short "one-liners" that inline well into the string fields of Kubernetes
API resources.

Inputs to a CEL program are "variables". Each Kubernetes API field that contains

> CEL declares in the API documentation which variables are available to use for

Reviewer 5: `contains CEL` -> `contains CEL expression`? How about we refer `cel` as the language and `cel expression` as the expression we used to validate through the whole doc?

Author: This reads OK to me, kinda like how one might say a field contains JSON.  Technically it might be more precise to say "contains a JSON object declaration" each time, but it's clear what is being stated.

> that field. For example, in the `x-kubernetes-validations.rules` field of

Reviewer 5: `x-kubernetes-validations.rules` -> `x-kubernetes-validations[i].rule`?

CustomResourceDefinitions, the `self` and `oldSelf` variables are available and
refer to the previous and current state of the custom resource data to be
validated by the CEL expression. Other Kubernetes API fields may declare
different variables. See the API documentation of the API fields to learn which
variables are available for that field.

Example CEL expressions:

| Rule                                                                               | Purpose                                                                           |
|------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------|
| `self.minReplicas <= self.replicas && self.replicas <= self.maxReplicas`           | Validate that the three fields defining replicas are ordered appropriately        |
| `'Available' in self.stateCounts`                                                  | Validate that an entry with the 'Available' key exists in a map                   |

> | `(self.list1.size() == 0) != self.list2.size() == 0)`                              | Validate that one of two lists is non-empty, but not both                         |

Reviewer 5: missing a `(` here

godoc for more information.

### Kubernetes Regex Library

In addition to the `matches` function provided by the CEL standard library, the
regex library provides `find` and `findAll`, enabling a much wider range of
regex operations.

Examples:

| CEL Expression                                              | Purpose                                                  |
|-------------------------------------------------------------|----------------------------------------------------------|
| `"abc 123".find('[0-9]*')`                                  | Find the first number in a string                        |
| `"1, 2, 3, 4".findAll('[0-9]*').map(x, int(x)).sum() < 100` | Verify that the numbers in a string sum to less than 100 |

See the [Kubernetes Regex Library](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library#Regex)
godoc for more information.

### Kubernetes URL Library

To make it easier and safer to process URLs, the following functions have been added:

- `isURL(string)` - Check if a string is a valid URL according to the [Go's
  net/url](https://pkg.go.dev/net/url#URL) package. The string must be an
  absolute URL.

> - `url(string) URL` Converts a string to a URL or results in an error if the

Reviewer 5: nit: `url(string) URL` Converts -> `url(string) URL - Convert` to keep style same as above `isURL(string)`.
Could also update `isURL(string) - Check` to `isURL(string) checks`

Author: Good catch. I resolved this the otherway.

  string is not a valid URL.

Once parsed via the `url` function, the resulting URL object has `getScheme`,
`getHost`, `getHostname`, `getPort`, `getEscapedPath` and `getQuery` accessors.

Examples:

| CEL Expression                                                  | Purpose                                        |
|-----------------------------------------------------------------|------------------------------------------------|
| `url('https://example.com:80/').getHost()`                      | Get the 'example.com:80' host part of the URL. |
| `url('https://example.com/path with spaces/').getEscapedPath()` | Returns '/path%20with%20spaces/'               |

See the [Kubernetes URL Library](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library#URLs)
godoc for more information.

### Kubernetes Authorization Library

> 

Reviewer 5: Note: This is going to be available starting 1.27 right? I noticed this pr is against `main` which means it will be available as soon as merged. 
I am ok with it since it mentioned checking API doc for `authorizer` variable just a note.

Author: Since this doesn't reference available particular capabilities in particular APIs, I don't see this as being particularly version sensitive. See below guidance about merging to main first.

> ### Kubernetes Authorization Library

Reviewer 3: I think it'll confuse people if this reference is in the v1.26 docs but an `authorizer` variable is never available with any CEL context in v1.26 clusters. _Someone_ will file an issue, is my guess.

I recommend saving this section for the v1.27 edition of this reference page.

> For CEL expressions where an `authorizer` variable is avaiable, the authorizer

Reviewer 3: (nit)
```suggestion
For CEL expressions where an `authorizer` variable is available, the authorizer
```

| CEL Expression                                                                         | Purpose                                                                                                               |
|----------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| `authorizer.group('').resource('pods').namespace('default').check('create').allowed()` | Check if the principal (user or service account) submitting a request is authorized to create pods in the default namespace.                         |
| `authorizer.path('/healthz').check('get').allowed()`                                   | Check if the principal (user or service account) submitting the request is authorized to to perform a HTTP GET for the non-resource `/healthz` path. |
| `authorizer.requestResource.check('custom-verb').allowed()`                            | Checks if the principal (user or service account) submitting the request is authorized for `custom-verb` on the resource being handled.              |

> See the [Authorization Library godoc](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library#Authz) for more information.

Reviewer 5: Also, this link is invalid til 1.27. Should we merge this PR to dev-1.27?

## Type Checking

CEL is a [gradually typed language](https://github.com/google/cel-spec/blob/master/doc/langdef.md#gradual-type-checking).

Some Kubernetes API fields contain fully type checked CEL expressions. For
example, [CustomResourceDefinitions Validation
Rules](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)
are fully type checked.

> Some Kubernetes API fields contain partially type checked CEL expressions. A

Reviewer 5: This is also available after 1.27

partially type checked expression is an experessions where some of the variables
are statically typed but others are dynamically typed. For example, in the CEL
expressions of
[ValidatingAdmissionPolicies](/docs/reference/access-authn-authz/validating-admission-policy/)
the `request` variable is typed, but the `object` variable is dynamically typed.
As a result, an expression containing `request.namex` would fail type checking
because the `namex` field is not defined. However, `object.namex` would pass
type checking even when the `namex` field is not defined for the resource kinds

> that `object` refers to, because `object` is dynamically typed.

Reviewer 5: Do we wanna mention that type checking failed here will not affect validation? I guess that is the info user will care.

Author: The only time type checking failures do not effect validation is when failurePolicy=ignore, but this paragraph is not focusing on that. I believe the way I've phrased this covers the actual behavior correctly. Let me know if I'm missing something.

| `string`      | `self.startsWith('kube')`         |

## Resource Constraints

CEL is non-Turing complete and offers a variety of production safety controls to
limit execution time. CEL's _resource constraint_ features provide feedback to
developers about expression complexity and help protect the API server from
excessive resource consumption during evaluation. CEL's resource constraint
features are used to prevent CEL evaluation from consuming excessive API server
resources.

A key element of the resource constraint features is a _cost unit_ that CEL
defines as a way of tracking CPU utilization. Cost units are independant of
system load and hardware. Cost units are also deterministic; for any given CEL
expression and input data, evaluation of the expression by the CEL interpreter
will always result in the same cost.

Many of CEL's core operations have fixed costs. The simplest operations, such as
comparisons (e.g. `<`) have a cost of 1. Some have a higher fixed cost, for
example:

- List literal declarations have a fixed base cost of 40 cost units.
- Authorization checks have a high fixed cost of 100000 cost units, effectively

>   limiting the total number of checks for any expression to 10.

Reviewer 3: If omitting authz, omit this too.
